lib/jormungandr/src/Cardano/Wallet/Jormungandr/Compatibility.hs

```diff
@@ -28,3 +51,81 @@ instance TxId Jormungandr where
 
 instance KeyToAddress Jormungandr where
     keyToAddress = undefined
+
+-- | Encode an 'Address' to a human-readable format. This produces two kinds of
+-- encodings:
+--
+-- - [Base58](https://en.wikipedia.org/wiki/Base58)
+--   for legacy / Byron addresses
+-- - [Bech32](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki)
+--   for Shelley addresses
+--
+-- The right encoding is picked by looking at the raw 'Address' representation
+-- in order to figure out to which class the address belongs.
+instance EncodeAddress Jormungandr where
+    encodeAddress _ (Address bytes) = do
+        if isJust (decodeLegacyAddress bytes) then base58 else bech32
+      where
+        base58 = T.decodeUtf8 $ encodeBase58 bitcoinAlphabet bytes
+        bech32 = Bech32.encodeLenient hrp (dataPartFromBytes bytes)
+
+-- | Decode text string into an 'Address'. Jörmungandr recognizes two kind of
+-- addresses:
+--
+-- - Legacy / Byron addresses encoded as `Base58`
+-- - Shelley addresses, encoded as `Bech32`
+--
+-- See also 'EncodeAddress Jormungandr'
+instance DecodeAddress Jormungandr where
+    decodeAddress _ x =
+        case (tryBech32, tryBase58) of
+            (Just bytes, _) -> bech32 bytes
+            (_, Just bytes) -> base58 bytes
+            (Nothing, Nothing) -> Left $ TextDecodingError
+                "Unable to decode address: encoding is neither Bech32 nor \
+                \Base58."
+      where
+        -- | Attempt decoding a legacy 'Address' using a Base58 encoding.
+        tryBase58 :: Maybe ByteString
+        tryBase58 = decodeBase58 bitcoinAlphabet (T.encodeUtf8 x)
+
+        -- | Verify the structure of a payload decoded from a Base58 text string
+        base58 :: ByteString -> Either TextDecodingError Address
+        base58 bytes = maybe (Left $ TextDecodingError errByron) Right $
+            decodeLegacyAddress bytes
+          where
+            errByron =
+                "Unable to decode address: neither Bech32-encoded nor a valid \
+                \Byron address."
+
+        -- | Attempt decoding an 'Address' using a Bech32 encoding.
+        tryBech32 :: Maybe (HumanReadablePart, ByteString)
+        tryBech32 = do
+            (hrp', dp) <- either (const Nothing) Just (Bech32.decodeLenient x)
+            (hrp',) <$> dataPartToBytes dp
+
+        -- | Verify the structure of a payload decoded from a Bech32 text string
+        bech32
+            :: (HumanReadablePart, ByteString)
+            -> Either TextDecodingError Address
+        bech32 (hrp', bytes) = do
+            when (hrp /= hrp') $ Left $ TextDecodingError $
+                "This address belongs to another network. Network is: "
+                <> show network <> "."
+            case BS.length bytes of
+                33 -> when (BS.take 1 bytes /= BS.pack [single]) $
+                    Left (invalidFirstByte single)
+                65 -> when (BS.take 1 bytes /= BS.pack [grouped]) $
```

Comment:
What is special about the constants `33` and `65`? Can they be given names (and a comment), instead of hard-coding them?

Reply:
1 discriminant byte + 32 bytes of spending key (+ 32 bytes of delegation key for grouped address). 

Named variables would be better indeed :+1:

Reply:
Fixed.